CI: Move Ubuntu 20.04 workflows to 22.04

franzpoeschel opened the pull request:
Ubuntu 20.04 runners shutting down April 1st

franzpoeschel committed:
Move Ubuntu 20.04 workflows to 22.04

franzpoeschel committed:
Upgrade linux and tooling runs to newer compiler versions

franzpoeschel committed:
Use Cuda PPA for Ubuntu 22.04

franzpoeschel committed:
Raise spack to 0.23.1

franzpoeschel committed:
Fix clangsanitizer

franzpoeschel committed:
Update spack envs

franzpoeschel committed:
Revert "Raise spack to 0.23.1"

This reverts commit c290fc6e8a7cc049b1aa997705243f91a0eb8a5c.

franzpoeschel committed:
Fixes

franzpoeschel committed:
Use mpicc/mpic++ in MPICH run

franzpoeschel commented on .github/ci/spack-envs/clang12_py38_mpich_h5_ad2/spack.yaml:
```diff
@@ -12,47 +12,47 @@ spack:
   compilers:
   - compiler:
-      environment: {}
+      environment: {MPICH_CC: /usr/lib/llvm-12/bin/clang, MPICH_CXX: /usr/lib/llvm-12/bin/clang++}
```
```suggestion
      environment: {}
```

franzpoeschel committed:
Update .github/ci/spack-envs/clang12_py38_mpich_h5_ad2/spack.yaml

franzpoeschel committed:
You will accept clang as your MPI compiler

if you want it or not

franzpoeschel committed:
Remove debugging output from previous commit

franzpoeschel committed:
Try sth

franzpoeschel committed:
Fix clang-tidy

franzpoeschel committed:
Revert "Try sth"

This reverts commit 672570fe4d0c08673c72c0e8894faa914e857be5.

franzpoeschel committed:
Try sth else

franzpoeschel committed:
Right place to disable clang-tidy checks

franzpoeschel committed:
Try requiring MPI

franzpoeschel committed:
????????????

franzpoeschel committed:
?

franzpoeschel committed:
We sneaky now

franzpoeschel committed:
asdf

franzpoeschel committed:
i dont get it

franzpoeschel commented:
@ax3l I can't get Spack to build with MPICH in clang12_py38_mpich_h5_ad2
If nothing else works, I'll go for a manual build, but maybe you have an idea

ax3l commented on .github/ci/spack-envs/clang12_py38_mpich_h5_ad2/spack.yaml:
```diff
@@ -12,47 +12,47 @@ spack:
-      spec: clang@8.0.0
+        cc: /usr/lib/llvm-12/bin/clang
+        cxx: /usr/lib/llvm-12/bin/clang++
+        f77: /usr/bin/mpif77
+        fc: /usr/bin/mpifort
```
```suggestion
        f77: /usr/bin/gfortran
        fc: /usr/bin/gfortran
```

ax3l commented on .github/workflows/linux.yml:
```diff
@@ -160,28 +160,42 @@ jobs:
+        sudo chmod +x /usr/bin/mpic++_clang
     - name: Build
-      env: {CC: clang-8, CXX: clang++-8, CXXFLAGS: -Werror}
+      env: {CC: mpicc_clang, CXX: mpic++_clang, CXXFLAGS: -Werror}
```
```suggestion
      env: {CC: clang++-12, CXX: clang++-12, CXXFLAGS: -Werror}
```

ax3l committed:
Undo workarounds clang12

ax3l committed:
Spack Clang12: `hdf5 +mpi`

For some reason concretized to `hdf5 ~mpi`, which is not Spack
default. Force MPI.

ax3l committed:
Remove More Clang 12 Work-Arounds

ax3l commented on .github/ci/spack-envs/clang12_py38_mpich_h5_ad2/spack.yaml:
```diff
@@ -12,47 +12,49 @@ spack:
       paths:
-        cc: /usr/lib/llvm-8/bin/clang
-        cxx: /usr/lib/llvm-8/bin/clang++
+        cc: /usr/lib/llvm-12/bin/clang
+        cxx: /usr/lib/llvm-12/bin/clang++
         f77: /usr/bin/gfortran
```
Not sure this path exists

ax3l replied:
fixed, it's `/usr/bin/clang-12` now

ax3l committed:
Fix Clang-12 Path

ax3l commented on .github/ci/spack-envs/clang11_nopy_ompi_h5_ad2/spack.yaml:
```diff
@@ -12,47 +12,47 @@ spack:
       paths:
-        cc: /usr/lib/llvm-7/bin/clang
-        cxx: /usr/lib/llvm-7/bin/clang++
+        cc: /usr/lib/llvm-11/bin/clang
+        cxx: /usr/lib/llvm-11/bin/clang++
```
```suggestion
        cc: /usr/bin/clang-11
        cxx: /usr/bin/clang++-11
```

ax3l commented on .github/ci/spack-envs/clang15_py311_nompi_h5_ad2/spack.yaml:
```diff
@@ -44,11 +44,11 @@ spack:
       paths:
-        cc: /usr/lib/llvm-14/bin/clang
-        cxx: /usr/lib/llvm-14/bin/clang++
+        cc: /usr/lib/llvm-15/bin/clang
+        cxx: /usr/lib/llvm-15/bin/clang++
```
```suggestion
        cc: /usr/bin/clang-15
        cxx: /usr/bin/clang++-15
```

ax3l commented on .github/ci/spack-envs/clangtidy_nopy_ompi_h5_ad2/spack.yaml:
```diff
@@ -12,47 +12,47 @@ spack:
       paths:
-        cc: /usr/lib/llvm-10/bin/clang
-        cxx: /usr/lib/llvm-10/bin/clang++
+        cc: /usr/lib/llvm-14/bin/clang
+        cxx: /usr/lib/llvm-14/bin/clang++
```
```suggestion
        cc: /usr/bin/clang-14
        cxx: /usr/bin/clang++-14
```

ax3l committed:
Update Two More Clang Paths

ax3l committed:
Fix Typo: clang-12

ax3l committed:
Remove OpenMPI? Default on system?

ax3l committed:
Set MPICH_CC/CXX

ax3l commented:
Wow, the last change definitely worked for my local docker attempt...

ax3l committed:
Remove `-Werror`

ax3l committed:
Modernize CTest: Use `--test-dir`

ax3l committed:
Spack Spec: Add `+mpi`

ax3l commented:
Oh, there is an MPICH bug marked as conflict for the exact MPICH and HDF5 combination we have here.
https://github.com/spack/spack/blob/c79b6207e88b765f9f01401f8145c17746457d09/var/spack/repos/builtin/packages/hdf5/package.py#L156

Let's go to Ubuntu 24.04 then for this runner?

ax3l committed:
Bump MPICH Runner to Ubuntu 24.04

ax3l commented:
I will remove ICC as well with this PR, since it is EOL for 1.5 years or so :)
ICX is the new compiler.

ax3l committed:
Remove ICC

ax3l committed:
More Clang14

ax3l committed:
Clang -> GCC

MPICH alone seems hard enough

ax3l committed:
GH: 4 vcores on Linux Now

franzpoeschel commented:
> Wow, the last change definitely worked for my local docker attempt...

I feel your pain

ax3l commented:
I simplified the test now to only test MPICH with GCC.

Currently it seems to work but the ParallelIO test fails with:
```
[Fri Mar 21 08:10:04 2025] [ADIOS2 EXCEPTION] <Toolkit>
  <format::BP5Deserializer> <InstallAttributeData> : Internal error or file
  corruption, no know format for Attribute Block
```
which I am not sure if it is a new MPICH issue or just a compiler build stack issue that we still need to solve in CI (more likely?)

ax3l commented:
In doubt, if we cannot fix this today, we can exlcude this one runner and add a separate PR that adds again a regular system GCC + system MPICH test to have it (MPICH) covered.

franzpoeschel commented:
Thank you for fixing the MPICH setup!
I don't think I ever saw this error. I can try investigating that next week, so let's maybe exlude the runner for the moment.

ax3l committed:
CI: Disable MPICH for now

ax3l commented:
Sounds good! Can you open a PR after this is merged that simply undo's the last commit I added here? :)

ax3l commented:
Yeah MPICH coverage is pretty important because we use it on all Cray/HPE systems (Frontier, Perlmutter, etc.) So let us definitely try to add one or two versions we can cover for it (and be aware of the HDF5 with MPICH 4.0 issues that Spack documents, no need to support that combination then).
